Ver1/Implemented and override ArrayList methods

NaKraVon opened the pull request.

mr-Edoc committed:
Ver1/Implemented and override ArrayList methods

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,130 @@
 package core.basesyntax;
 
+import java.util.Arrays;
```
**Violation — banned utility imports used.**

This file imports `java.util.Arrays` and `java.util.Objects`, which violates the checklist item: **"Don't use Objects, Arrays, or any other util class."** Remove these imports and avoid using their methods. For example, replace `Objects.equals(a, b)` with a null-safe comparison (`a == null ? b == null : a.equals(b)`), and replace `Arrays.copyOf(...)` with explicit allocation and `System.arraycopy(...)`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,130 @@
+            if (newCapacity < data.length) { // Check for overflow
+                newCapacity = Integer.MAX_VALUE;
+            }
+            this.data = Arrays.copyOf(data, newCapacity);
```
**Violation — using Arrays.copyOf in addAll; must use System.arraycopy().**

At line 49 the code calls `Arrays.copyOf(data, newCapacity)`. This breaks the checklist requirement: **"Use System.arraycopy() to move your array elements."** Allocate a new array of the required capacity and use `System.arraycopy(data, 0, newArray, 0, size)` to copy elements instead of `Arrays.copyOf`.

Also note this use of `Arrays.copyOf` is a direct example of using the banned util class from: **"Don't use Objects, Arrays, or any other util class."**

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,130 @@
+            if (Objects.equals(element, data[i])) {
+                for (int j = i + 1; j < size; j++) {
+                    data[j - 1] = data[j];
+                }
```
**Violation — manual element shifting and use of Objects.equals in remove(T).**

Lines 86–89 use `Objects.equals(element, data[i])` and then shift elements with a manual loop. This violates two checklist items:
- **"Don't use Objects, Arrays, or any other util class."** (no `Objects.equals`)
- **"Use System.arraycopy() to move your array elements."** (no manual loop for shifting)

Replace `Objects.equals(...)` with a null-safe `equals` check and use `System.arraycopy(data, i + 1, data, i, size - i - 1)` to shift the tail left by one, then decrement `size` and null out the freed slot.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,130 @@
+        if (newCapacity < data.length) { // Check for overflow
+            newCapacity = Integer.MAX_VALUE;
+        }
+        this.data = Arrays.copyOf(data, newCapacity);
```
**Violation — grow() uses Arrays.copyOf and lacks named growth-factor constant.**

The `grow()` method (lines 107–112) computes `int newCapacity = data.length + (data.length >> 1);` (line 108) and then calls `Arrays.copyOf` (line 112). This violates:
- **"Any magic numbers should be constants"** — the 1.5x growth should be represented by a named constant (for example `GROWTH_FACTOR_NUMERATOR = 3` and `GROWTH_FACTOR_DENOMINATOR = 2`, or `GROWTH_FACTOR = 1.5`).
- **"Use System.arraycopy() to move your array elements."** — don't use `Arrays.copyOf`.

Refactor to compute the new capacity using a descriptive constant and to allocate a new array and copy elements with `System.arraycopy(data, 0, newArray, 0, size)`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,130 @@
+        checkRangeForAdd(index);
+
+        if (size == data.length) {
+            grow();
```
**Violation — method name `grow()` is ambiguous; use more informative name.**

The checklist requests informative method names: **"Use informative names for your variables and methods."** The method `grow()` (declared at line 107) and its calls at lines 21 and 32 would be clearer if named `growIfArrayFull()` or `ensureCapacityForAdd()` to indicate the conditional nature. Consider renaming the method and keeping its responsibility focused on resizing only (single responsibility).

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision: Request changes**

Your ArrayList implementation covers core behaviors well, but it violates several explicit checklist requirements that are considered blockers.

Critical issues to fix:
- Banned utilities: remove imports/usages of `java.util.Arrays` and `java.util.Objects` (e.g., `Objects.equals` at lines 86–89 and `Arrays.copyOf` at line 49 and in `grow()`), per “Don’t use Objects, Arrays, or any other util class.” Replace equality with a null-safe check: `a == null ? b == null : a.equals(b)`.
- Use `System.arraycopy()` for all element moves/copies. In `addAll` (around line 49) and `grow()` (lines 107–112), allocate a new array and copy via `System.arraycopy(data, 0, newArr, 0, size)`. In `remove(T)` (lines 86–89), shift with `System.arraycopy(data, i + 1, data, i, size - i - 1)`, then `size--` and null the freed slot.
- Replace magic growth math with a named constant. Introduce a descriptive growth factor constant and compute new capacity using it, rather than `data.length + (data.length >> 1)`.
- Centralize resizing. Replace ad-hoc logic in `addAll` with a single `ensureCapacity(minCapacity)` helper that performs the allocation + `arraycopy`.
- Naming: rename `grow()` to something more informative like `ensureCapacity` or `growIfArrayFull`.

Minor: ensure there’s no duplicate shifting logic and keep whitespace tidy. You’re close—addressing these will bring the code in line with the checklist. Keep it up!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

mr-Edoc committed:
Ver1.1/Fixed few methods according to review
Change grow() name to growIfArrayFull(), remove Object and Arrays import classes, fix addAll logic, switch if in remove(T element) to null safe check.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,130 @@
+            if (element == null ? data[i] == null
+                    : element.equals(data[i])) {
+                for (int j = i + 1; j < size; j++) {
+                    data[j - 1] = data[j];
```
This violates checklist item: "Use [System.arraycopy()](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#arraycopy-java.lang.Object-int-java.lang.Object-int-int-) to move your array elements." In `remove(T element)` the code shifts elements using a manual for-loop (`for (int j = i + 1; j < size; j++) { data[j - 1] = data[j]; }`). Replace the manual element shift with `System.arraycopy(data, i + 1, data, i, size - i - 1);` then decrement `size` and null the freed slot to match the rest of the implementation and the checklist guidance.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,130 @@
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int DIVIDED_BY_TWO = 1;
```
This violates checklist items: "Any magic numbers should be constants" and "Use informative names for your variables and methods." The constant `DIVIDED_BY_TWO` (value `1`) is not an informative name describing growth factor or its purpose. The requirement expects a clear, descriptive constant for the growth logic (for example `GROW_SHIFT` or an explicit `GROWTH_FACTOR` constant). Rename/rework this so the growth purpose is obvious to readers and maintainers.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,130 @@
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int DIVIDED_BY_TWO = 1;
+    private static final int GROW_COPY_INDEX = 0;
```
This violates checklist item: "Don't create redundant variables." The constant `GROW_COPY_INDEX = 0` is redundant and only used once in the arraycopy call. Remove this constant and use the literal `0` in the copy call (or otherwise justify the constant with a clear name and usage). Keeping redundant constants reduces clarity.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Requesting changes because there are checklist violations that are considered critical for this task’s acceptance.

Critical issues to fix:
- At `src/main/java/core/basesyntax/ArrayList.java`, in `remove(T element)`, you manually shift elements with a `for` loop (around the block after a match is found). This violates the requirement to use `System.arraycopy` for element moves. Replace the loop with `System.arraycopy(data, i + 1, data, i, size - i - 1);`, then do `size--` and set `data[size] = null;` to avoid memory leaks and return the removed stored value (store it in a local before shifting). [CHECKLIST ITEM]
- The constant naming for growth is unclear: `DIVIDED_BY_TWO = 1` does not describe its purpose and reads as a magic number. Rename/rework this into a clear growth-related constant (e.g., `GROWTH_FACTOR_NUM = 3`, `GROWTH_FACTOR_DEN = 2`, or a descriptive `GROWTH_FACTOR`) and use it where capacity is calculated. [CHECKLIST ITEM]
- Remove the redundant `GROW_COPY_INDEX = 0` constant and use the literal `0` directly in `System.arraycopy`, or justify it with a more meaningful name and multiple uses. [CHECKLIST ITEM]

Minor suggestions (non-blockers): consider copying only `size` elements during grow, not the whole buffer length; also ensure formatting has no leading blank lines.

Overall, great job on default capacity, growth behavior, centralized index checks, and exception handling. Addressing the above will align your solution fully with the checklist and improve clarity.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

mr-Edoc committed:
Ver1.2/Fixed few methods according to review
Change DIVIDED_BY_TWO to GROWTH_FACTOR, remove for loop in remove(T element) and amke it System.arraycopy instead, remove GROW_COPY_INDEX.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mr-Edoc committed:
Ver1.2/Fixed few methods according to review
Change DIVIDED_BY_TWO to GROWTH_FACTOR, remove for loop in remove(T element) and amke it System.arraycopy instead, remove GROW_COPY_INDEX.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Great work overall—your ArrayList meets the key requirements: default capacity of 10, 1.5× growth, centralized index checks, clear exceptions, and consistent use of System.arraycopy. 

> Educational override: Approving to help you progress while learning; please address the following for correctness.

- Critical: In `remove(T element)` (around the return statement after shifting), you return the method parameter instead of the actual stored value. This can return a different instance than what was in the list, which breaks expected semantics. Fix by storing the found value first (e.g., `T oldValue = data[i];`), perform the shift, null out the tail, then `return oldValue`.
- Minor: In your grow method, copy only `size` elements instead of the entire buffer to avoid unnecessary work. This improves efficiency without changing behavior.

Strengths to keep: clear single-responsibility methods, informative error messages, and correct index-bound checks. Nice job aligning behavior with java.util.ArrayList expectations.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
